child/child.go

```diff
@@ -0,0 +1,345 @@
+package child
+
+import (
+	"errors"
+	"io"
+	"log"
+	"math/rand"
+	"os"
+	"os/exec"
+	"sync"
+	"syscall"
+	"time"
+)
+
+var (
+	// ErrMissingCommand is the error returned when no command is specified
+	// to run.
+	ErrMissingCommand error = errors.New("missing command")
+
+	// ExitCodeOK is the default OK exit code.
+	ExitCodeOK int = 0
+
+	// ExitCodeError is the default error code returned when the child exits with
+	// an error without a more specific code.
+	ExitCodeError int = 127
+)
+
+// Child is a wrapper around a child process which can be used to send signals
+// and manage the processes' lifecycle.
+type Child struct {
+	sync.RWMutex
+
+	stdin          io.Reader
+	stdout, stderr io.Writer
+	command        string
+	args           []string
+
+	reloadSignal os.Signal
+
+	killSignal  os.Signal
+	killTimeout time.Duration
+
+	splay time.Duration
+
+	// cmd is the actual child process under management.
+	cmd *exec.Cmd
+
+	// exitCh is the channel where the processes exit will be returned.
+	exitCh chan int
+
+	// stopLock is the mutext to lock when stopping. stopCh is the circuit breaker
+	// to force-terminate any waiting splays to kill the process now. stopped is
+	// a boolean that tells us if we have previously been stopped.
+	stopLock sync.RWMutex
+	stopCh   chan struct{}
+	stopped  bool
+}
+
+// NewInput is input to the NewChild function.
+type NewInput struct {
+	// Stdin is the io.Reader where input will come from. This is sent directly to
+	// the child process. Stdout and Stderr represent the io.Writer objects where
+	// the child process will send output and errorput.
+	Stdin          io.Reader
+	Stdout, Stderr io.Writer
+
+	// Command is the name of the command to execute. Args are the list of
+	// arguments to pass when starting the command.
+	Command string
+	Args    []string
+
+	// ReloadSignal is the signal to send to reload this process. This value may
+	// be nil.
+	ReloadSignal os.Signal
+
+	// KillSignal is the signal to send to gracefully kill this process. This
+	// value may be nil.
+	KillSignal os.Signal
+
+	// KillTimeout is the amount of time to wait for the process to gracefully
+	// terminate before force-killing.
+	KillTimeout time.Duration
+
+	// Splay is the maximum random amount of time to wait before sending signals.
+	// This option helps reduce the thundering herd problem by effectively
+	// sleeping for a random amount of time before sending the signal. This
+	// prevents multiple processes from all signaling at the same time. This value
+	// may be zero (which disables the splay entirely).
+	Splay time.Duration
+}
+
+// New creates a new child process for management with high-level APIs for
+// sending signals to the child process, restarting the child process, and
+// gracefully terminating the child process.
+func New(i *NewInput) (*Child, error) {
+	if i == nil {
```

Comment:
If they don't specify anything, the check that follows will always fail, so you can delete this or just make sure `i` isn't `nil`.